csrc/root_domain_map.cpp

```diff
@@ -104,61 +120,70 @@ std::unordered_map<IterDomain*, IterDomain*> PairwiseRootDomainMap::map(
     squeeze_flags = sop->getSqueezeDimFlags();
   }
 
-  IterDomain* selected_id = nullptr;
-  bool select_skip_consumer = true;
-  if (SelectOp* sop = dynamic_cast<SelectOp*>(consumer_tv_->definition())) {
-    selected_id = sop->getSelectAxis();
-    select_skip_consumer = false;
-  } else if (
-      IndexSelectOp* sop =
-          dynamic_cast<IndexSelectOp*>(consumer_tv_->definition())) {
-    selected_id = sop->getSelectAxis();
-    select_skip_consumer = true;
-  } else if (
-      TorchGatherOp* gop =
-          dynamic_cast<TorchGatherOp*>(consumer_tv_->definition())) {
-    selected_id = gop->getSelectAxis();
-    select_skip_consumer = true;
-  }
+  auto [indexed_producer_id, has_consumer_of_indexed_id] =
+      getIndexedDomainInfo(producer_tv_, consumer_tv_);
 
   std::unordered_map<IterDomain*, IterDomain*> dom_map;
   const auto producer_root =
       TensorDomain::noReductions(producer->getMaybeRFactorDomain());
   const auto& consumer_root = consumer->getRootDomain();
   size_t itc = 0, itp = 0;
   while (itc < consumer_root.size() && itp < producer_root.size()) {
-    IterDomain* producer_id = producer_root[itp];
-    IterDomain* consumer_id = consumer_root[itc];
-
-    // When the producer ID is the dim of a select-like op, there is no
-    // mapping for it.
-    if (producer_id == selected_id) {
-      itp++;
-      if (select_skip_consumer) {
+    IterDomain* producer_id = producer_root.at(itp);
+    IterDomain* consumer_id = consumer_root.at(itc);
+
+    // Conditions to check:
+    // 1. Indirectly accessed IDs (e.g., select)
+    // 2. IDs that may have different extents (e.g., non indexed
+    //  domains of torch_gather)
+    // 3. Squeeze and unsqueeze
+    // 4. Broadcast and non broadcast
+
+    // Condition 1: when the producer ID is the dim of a select-like op
+    if (producer_id == indexed_producer_id) {
+      // If there's no corresponding consumer, skip the indexed producer
+      if (!has_consumer_of_indexed_id) {
+        itp++;
+        continue;
+      }
+      // Skip both producer and consumer if mapping not allowed
+      if (!map_indexed_domains_) {
+        itp++;
         itc++;
+        continue;
       }
+    }
+
+    // Condition 2: Different extents
+    if (consumer_tv_->definition()->isA<TorchGatherOp>() &&
```

Comment:
nit: in C++17, you can do
```
if (auto gop = dynamic_cast<TorchGatherOp*>(consumer_tv_->definition();  gop != nullptr && gop->exactSizes() && ...)
```
which is slightly less verbose.

Reply:
Ha, that's nice. 

I found this article helpful: https://skebanga.github.io/if-with-initializer/